### ISSUE TITLE
Moment.js as a peer dependency

### ISSUE THREAD

**Issue by arqex** [Hacker0x01/react-datepicker#122]
Hi,

Moment.js is a big dependency that I don't want to add twice to my bundle. Could it be moved to the peerDependencies in `package.json` like react?.

Thanks for this component, works great.

Cheers,
Javi

**Comment by martijnrusschen** [comment]
I tried moving the dependency to peerdependencies, but when running npm install I got the following message:

```
➜  react-datepicker git:(dependencies) npm install
npm WARN peerDependencies The peer dependency moment included from react-datepicker will no
npm WARN peerDependencies longer be automatically installed to fulfill the peerDependency 
npm WARN peerDependencies in npm 3+. Your application will need to depend on it explicitly.
```

Not sure if we want to change this or not, the documentation I found about the difference between dependencies is not really clear to me, but maybe you could help me out with this.

**Comment by arqex** [comment]
Hi @martijnrusschen 

A peer dependency is a module needed by the package to work but that it is not installed with the package. That is, if I need to use react-datepicker and moment.js is a peer dependency of it, I will need to have moment.js also installed to make it work, but react-datepicker won't install it for me.

In my case I am already using moment.js in my projects. Moment.js is a moderately big library that adds ~50KB to my bundle, if react-datepicker install it again. it will be repeated inside the bundle adding 50KB more to the final size.

I think moment.js will be better as a peerDependency than a regular dependency. React-datepicker already uses React as a peer dependency.

**Comment by RSO** [comment]
Hey @arqex,

Sorry for getting back to you this late, I've been away on vacation.

I agree that for now momentjs is better suited to be a `peerDependency`, but I'm not completely sure yet as to how that would work. Does it mean that I'll have to manually install momentjs in order to be able to develop on the datepicker? Or should I maybe make it a `devDependency` and a `peerDependency`?

npm/npm#6565 seems to state (like @martijnrusschen's comment) that in the future `peerDependencies` aren't installed automagically anymore, but when I run `npm install` it also doesn't seem to install them, so I'm really not sure how this all should work out.

I'll try and play with this a little further, have you tried getting this working on your own? If so, I would really appreciate a PR!

**Comment by RSO** [comment]
I'm also not quite sure where to draw the line between `dependencies` and `peerDependencies`, why not make all `dependencies` `peerDependencies`?

**Comment by RSO** [comment]
So, I did some more reading on `peerDependencies`, and based on https://github.com/npm/npm/issues/5080#issuecomment-40510098 I think that we shouldn't use `peerDependencies` for this. Could you maybe try and see if the the `npm dedupe` command fixes your problem?

**Comment by arqex** [comment]
Hi @RSO 

I think they won't be installed automatically as for npm v3.0 and ahead.

Wow! `npm dedupe`! I have never heard of it, I should have a look. About peerDependencies being deprecated, it is a tough matter. You will never make react a real dependency, as long as you don't want to break any react application installing react-datepicker, but it is true that is difficult to be sure when a dependency is a peerDependency.

In the case of moment.js, maybe the best option is to create two different builds, one with moment.js and other without it, like https://github.com/zippyui/react-date-picker/tree/master/dist

Last week I have been creating my own datepicker because I needed a timepicker too, and I left moment.js as peerDependency, and it is true that I need to install it besides the datepicker to make it work, but on the other hand I get a really lightweight build:

https://github.com/arqex/react-datetime/tree/master/dist

I think I will leave it as peerDependency but it looks like a matter of taste, isn't it?

**Comment by RSO** [comment]
How are you including the datepicker? Are you doing a `require('react-date-picker')` or are you include the `react-datepicker.js` file in your source? The latter shouldn't include momentjs.

I just checked the `react-datepicker.js` file and it seems odd that it still includes Tether and lodash, so I guess I'll remove those from the package as well, and assume that all dependencies are loaded externally.

**Comment by martijnrusschen** [comment]
@rafeememon What do you think? Should we change this? I checked our dependency list and it looks like we're relying on small lib mostly, but moment.js is the only heavy lib in this list. Should we move this (breaking change) or just keep is as it is?

**Comment by rafeememon** [comment]
I think this should be changed, but I think we should wait to break this right before the v1.0 release.  The reasoning behind this is that the DatePicker depends on moment, but it's extremely likely that consumers will be depending on moment independently in their app code (at the very least to use the DatePicker API), so it should use the same physical dependency to line up the versions.

**Comment by aij** [comment]
@martijnrusschen, @Kerumen It shouldn't break for *everyone* -- only for users who depend on npm pulling in `moment` transitively via `react-datepicker`. Do we know how common that is?

Those users could fall into two categories:

1. Users who never use/call `moment()` directly.
2. Users who do call `moment()` directly but failed to declare the dependency.

Group 1 is correct usage, which would unfortunately be broken by this change when they choose to upgrade. They would only be using react-datepicker in the simplest ways, since they couldn't be passing in any moment props without creating a moment.

Group 2 is  already rather fragile, since any change that makes `moment` no longer available project-wide would break their code. For example, a library upgrade or addition that introduces a new `moment` dependency on an incompatible version would already break those projects.

Either way, users who would be broken by switching to `peerDependencies` should get a warning when they upgrade. (Though the warning may be easy to miss/ignore...)

I happen to know the projects I'm working on wouldn't be broken by this change. (Though we are also way out of date anyway due to other breaking changes and lack of testing resources.)

**Comment by Kerumen** [comment]
I totally agree with @aij. I think it's a safe move.

**Comment by aij** [comment]
@Kerumen I didn't say it's *safe*, just that it won't break for everyone. I don't know how everyone is using the library, so I can't tell whether it will be a breaking change for a lot of people or only a few.

I would hope only a few though, and I would hope they would (be able to) see the warnings/errors from NPM, but I expect there would be at least some who don't.

**Comment by Kerumen** [comment]
Yes I get what you mean. We can warn in the changelog/releases/readme and I'd be happy to help people if they file an issue regarding their project :)

**Comment by martijnrusschen** [comment]
Thanks! Released this change in v0.64.0
